core/pta/widevine.c

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2023, The ChromiumOS Authors
+ */
+
+#include <compiler.h>
+#include <initcall.h>
+#include <kernel/boot.h>
+#include <kernel/dt.h>
+#include <kernel/pseudo_ta.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <mm/tee_mm.h>
+#include <mm/tee_pager.h>
+#include <stdio.h>
+#include <string.h>
+#include <string_ext.h>
+#include <tee_api.h>
+#include <trace.h>
+
+#define PTA_NAME "widevine.pta"
+
+#define GET_TPM_PUBKEY 0
+#define GET_WIDEVINE_PRIVKEY 1
+
+/* 721f4da9-da05-40d4-a1a3-8377c1e08b0a */
+#define WIDEVINE_PTA_UUID                                              \
+	{                                                              \
+		0x721f4da9, 0xda05, 0x40d4,                            \
+		{                                                      \
+			0xa1, 0xa3, 0x83, 0x77, 0xc1, 0xe0, 0x8b, 0x0a \
+		}                                                      \
+	}
+
+#define TPM_AUTH_PUB_MAX_SIZE 512
+#define WIDEVINE_PRIV_MAX_SIZE 32
+
+/* The public key to encrypt the communication to TPM. */
+static uint8_t tpm_auth_pub[TPM_AUTH_PUB_MAX_SIZE];
+static uint32_t tpm_auth_pub_size;
+
+/* The private key for Widevine signing. */
+static uint8_t widevine_priv[WIDEVINE_PRIV_MAX_SIZE];
+static uint32_t widevine_priv_size;
+
+static TEE_Result init_widevine_dt_data(void)
+{
+	int node = 0;
+	int len = 0;
+	void *fdt;
+	const void *result = NULL;
+
+	fdt = get_dt();
+
+	if (!fdt)
+		return TEE_ERROR_NO_DATA;
+
+	node = fdt_path_offset(fdt, "/options/widevine");
+	if (node < 0)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	result = fdt_getprop(fdt, node, "tcg-tpm-auth-public-key", &len);
+	if (!result)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len > TPM_AUTH_PUB_MAX_SIZE)
+		return TEE_ERROR_OVERFLOW;
+
+	memcpy(tpm_auth_pub, result, len);
+	tpm_auth_pub_size = len;
+
+	result = fdt_getprop(fdt, node, "widevine-root-of-trust", &len);
+	if (!result)
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	if (len > WIDEVINE_PRIV_MAX_SIZE)
+		return TEE_ERROR_OVERFLOW;
+
+	memcpy(widevine_priv, result, len);
+	widevine_priv_size = len;
```

Comment:
I am probably missing something obvious here, but, am I reading this correctly that we're reading the private key from device tree and then copy it into a secure (TEE) memory? If yes, how if this secure? What prevents the private key from being read out by others?